aprx-stat.8
.\" APRX-STAT  v.0.12
.TH aprx\-stat 8 "2007 December 25 \- v0.12"
.LO 8
.SH NAME
.B aprx\-stat
\- statistics utility for
.BR aprx (8)
.SH SYNOPSIS
aprx-stat [\-t] [\-f /tmp/aprx\-erlang.dat] {\-S|\-x|\-X}
.SH DESCRIPTION
.B aprx\-stat
is a statistics utility for
.BR aprx (8)
program.

.SH OPTIONS
The
.B aprx\-stat
has following runtime options:
.TP
.B "\-t"
Use UNIX
.I time_t
for timestamps, instead of human readable text format.
.TP
.B "\-f /tmp/aprx\-erlang.dat"
Turn on verbose debugging, outputs data to STDOUT.
.TP
.B "\-S"
SNMP data mode, current counter and gauge values.
.TP
.B "\-x"
Lattest of extended historical gauge values.
This gives for each input interface
.RS
.IP \(bu 2
SNMP data
.IP \(bu 2
last 90 of 1 minute values,
.IP \(bu 2
10 of 10 minute values,
.IP \(bu 2
3 of 60 minute values.
.RE
.TP
.B "\-X"
Full extended historical gauge values.
This gives all the contents of historical value data ring-buffers.
.RS
.IP \(bu 2
SNMP data
.IP \(bu 2
1 minute resolution: 24 hours
.IP \(bu 2
10 minute resolution: 7 days
.IP \(bu 2
60 minute resolution: 3 months
.RE

.SH SNMP DATA OUTPUT
For each interface feeding AX.25 packets and/or KISS frames to this program,
there are following kind of 
.nf
\fC
SNMP  /dev/ttyUSB1   798282 11088   0  0     3
SNMP  ax0   798282 11088   0  0     7
SNMP  ax1   798282 11088   0  0     94
.fi
.PP
where columns are:
.IP \(bu 2
"SNMP" - keyword
.IP \(bu 2
Interface (AX.25 IF name, or serial port device name)
.IP \(bu 2
Received byte counter
.IP \(bu 2
Received KISS frame (packet) counter
.IP \(bu 2
Transmitted byte counter (will stay zero)
.IP \(bu 2
Transmitted KISS frame counter  (will stay zero)
.IP \(bu 2
Age in seconds of last update of this statistics.

.SH EXTENDED DATA OUTPUT
Extended data output gives formatted historical periodic accumulates of interface traffic
counters, and Erlang value estimates based on that.
.PP
.nf
\fC

SNMP  /dev/ttyUSB1   816675 11332   0  0     15

1min data
2007-12-24 14:10  /dev/ttyUSB1  1m    374    6      0    0   0.047  0.000
2007-12-24 14:09  /dev/ttyUSB1  1m    377    5      0    0   0.047  0.000
2007-12-24 14:08  /dev/ttyUSB1  1m    347    5      0    0   0.043  0.000
2007-12-24 14:07  /dev/ttyUSB1  1m    140    2      0    0   0.018  0.000
\(bu\(bu\(bu

10min data
2007-12-24 14:10  /dev/ttyUSB1 10m   3829   55      0    0   0.048  0.000
2007-12-24 14:00  /dev/ttyUSB1 10m   2182   28      0    0   0.027  0.000
2007-12-24 13:50  /dev/ttyUSB1 10m   3205   44      0    0   0.040  0.000
2007-12-24 13:40  /dev/ttyUSB1 10m   3811   50      0    0   0.048  0.000
\(bu\(bu\(bu

60min data
2007-12-24 14:00  /dev/ttyUSB1 60m  22510  295      0    0   0.047  0.000
2007-12-24 13:00  /dev/ttyUSB1 60m  24886  347      0    0   0.052  0.000
\(bu\(bu\(bu
.fi
.PP
The output repeats for all interfaces.
.PP
The SNMP dataset is given in the beginning, and described above.
Then each extended output line has following fields:
.IP \(bu 2
Timestamp is two fields, date and time (in minute resolution) is in UTC.
.IP \(bu 2
Alternate timestamp format is UNIX
.I time_t
as an integer, counting seconds from epoch, and as single field.
.IP \(bu 2
Interface name is same as in SNMP case.
.IP \(bu 2
Data qualifier tells what integration period the data is valid for:
.IR 1m ", " 10m ", " 60m .
.IP \(bu 2
Counter of received bytes on interface (including KISS flags etc.)
.IP \(bu 2
Counter of received frames.
.IP \(bu 2
Counter of transmitted bytes on interface
.IP \(bu 2
Counter of transmitted frames.
.IP \(bu 2
Reception
.I Erlang 
value estimate.
.IP \(bu 2
Transmission
.I Erlang
value estimate.
.PP
.SH TODO
.SH BUGS
.SH SEE ALSO
.BR aprx (8)

.SH CONFIGURATION FILE
There is no configuration file.

.SH NOTES: ERLANG
The
.I Erlang
is telecom measurement of channel occupancy, and in this application sense
it does tell how much traffic there is on the radio channel.
.PP
Most radio transmitters are not aware of all transmitters on channel,
and thus there can happen a collision causing loss of both messages.
The higher the channel activity, the more likely that collision is.
For further details, refer to statistical mathematics books, or perhaps
on Wikipedia.
.PP
In order to measure channel activity, the
.B aprx
program suite has these built-in statistics counter and summary estimators.
.PP
The
.I Erlag
value that the estimators present are likely somewhat
.I underestimating
the true channel occupancy simply because it calculates estimate of channel
bit transmit rate, and thus a per-minute character capacity.
It does not know true frequency of bit-stuffing events of the HDLC framing,
nor each transmitter pre- and port frame PTT times. The transmitters need to
stabilize their transmit oscillators in many cases, which may take up to
around 500 ms!
The counters are not aware of this preamble-, nor postamble-times.
.PP
The HDLC bit stuffing ratio is guessed to be 8.2 bits for each 8 bits of payload.


.SH NOTES: SUID ROOT
This program needs probably to be run as
.I "suid\-root" 
!
It is fairly safe to do so, as this checks that the
.B "\-f"
parameter file is of correct time, and will not try to create create
it if it does not exist.

.SH AUTHOR
This little piece was written by
.I "Matti Aarnio, OH2MQK"
during a dark and rainy fall and winter of 2007\-2008 after a number
of discussions grumbling about current breed of available software
for APRS iGate use in Linux (or of any UNIX) platforms.
.PP
Principal contributors and test users include:
.IR "Pentti Gronlund, OH3BK" ,
.IR "Reijo Hakala, OH1GWK" .


ChangeLog
2008-01-05  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* Makefile:
	    Radical revisioning, Best Current Practice

	* aprx.c:
	    Write aprx.pid  file in all cases, not only when starting
	    as daemon.

	* version 0.13

	* aprx.h, aprx.c, erlang.c, aprx-stat.c:
	    Modifications on shared memory segment head, stores
	    MYCALL, correct running process PID.

	* INSTALL:
	    Some text fixes

	* aprx-stat.c, aprx.h, erlang.c:
	    Report also server process PID on the SNMP dataset,
	    and time in seconds since it was started.

	* ax25.c:
	    Verify TNC2 format APRS message's FROM>DEST,VIA,VIA:
	    callsigns to be of proper syntax.

	* Makefile, aprx.8, aprx.8.in, aprx-stat.8, aprx-stat.8.in,
	  aprx.conf, aprx.conf.in:
	    Centralized a bit of configuring into Makefile, generating
	    files from *.in versions.

	* aprxpolls.c:
	    Library function used everywhere, part of "eliminate fixed
	    size preallocations" -task.
	
	* aprsis.c:
	    Debug logging improvements, parent death detection.
	    Preserved MAXAPRSIS setting - of 10 servers.

	* ttyreader.c:
	    Unfix the number of TTYs, now can be as many as one wants.

	* config.c, aprx.c, aprx.h, aprx.conf.in, aprx.8.in:
	    pidfile configuration parameter, and its usage

2008-01-04  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* VERSION, arpx.c, Makefile, aprx.8, aprx.8.in,
	  aprx-stat.8, aprx-stat.8.in:
	    Define system version in VERSION file, have it stamped
	    on programs and 

	* config.c:
	    The "aprxlog" and "rflog" parameters were interchanged
	    at some point.

	* aprx.c, aprsis.c, beacon.c:
	    Complain loudly with -d or -v options on, and if the
	    configuration does not set global  mycall  parameter.
	    Do not refuse to run, though!  This is perfectly valid
	    for things like Erlang-monitoring.

2007-12-29  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* version aprx-v0.12

	* PROTOCOLS:
	    Writeup of existing (and planned) protocols that this
	    software uses.

	* ax25.c:
	    Correct processing of 3rd-party frames.
	    Also separate TNC2 formatted frame Rx-igateing rules
	    from AX.25-to-TNC2 format translation routine.

	* ax25.c, beacon.c, aprsis.c:
	    Centralize the APRSIS communication line ending CRLF char pair
	    addition into common code, not distributed all over.

2007-12-25  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* version aprx-v0.11

	* TODO, README:
	    Cleanup

	* config.c, netax25.c, aprx.8, aprx.conf:
	    Config option "ax25-rxport" - Limits acceptance of APRS
	    packets only from listed Linux AX.25 ports.

	* ttyreader.c, aprx.conf, aprx.8:
	    Reworked a bit of the serialport config options.
	    Now initstring, and various KISS modes are configurable

	* erlang.c:
	    Carefull approach on erlang-file opening, if file exists,
	    and is non-zero size, open it only if magics match.

	* config.c:
	    config_SKIPTEXT() terminates all scanned strings with NUL byte,
	    and moves to byte following it, if the termination byte was not
	    a NUL byte originally.

	* aprx-stat.c, aprx-stat.8:
	    Option -t  to show timestamps differently.

	* Makefile, man-to-html.sh:
	    Produce decent format HTML versions of the man-pages.

	* config.c, erlang.c, aprx.8, aprx.conf:
	    Added "erlang-log1min" option to control 1 minute interval
	    Erlang sampling logging behaviour.

	* Makefile:
	    Fix "make install" of man-pages

	* version aprx-v0.10

	* aprx-stat.c, aprx-stat.8, erlang.c, Makefile:
	    Statistics reporter tool.

	* aprsis.c, config.c, aprx.8, aprx.conf:
	    Multiple aprsis-server  config definitions are now
	    supported, and used in round-robin fashion.

	* ax25.c, config.c, aprx.8, aprx.conf:
	    Config option  ax25-filter

	* ttyreader.c, erlang.c:
	    Magic channel capacity constant expressions updated..

2007-12-23  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprx.c, aprx.h, aprsis.c, aprx.8, aprx.conf, ax25.c, erlang.c:
	    Multiple configuration file options, rf-log, and aprx-log -files,
	    related documentation.

	* config.c, aprx.8:
	    Special quoted-string escape processing.

	* version aprx-v0.08

	* erlang.c, aprx.c, aprx.h:
	    Erlang data has now a mmap():ed filesystem based backing-store.
	    Erlang-data can be syslog()ed, and independently of that,
	    it can be printed on STDOUT.  Default syslog facility is "NONE".
	
	* Great Rename -- the thing is now called:  aprx

2007-12-06  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprsg.8, erlang.c, aprsg.c, netax25.c, ttyreader.c:
	    Erlang logging now uses syslog(3), unless explicitely told to use
	    output to stdout.  Also the Erlang log format was altered a bit,
	    now it reports also number of packets in the interval.

	* erlang.c, Makefile:
	    "make ERLANG1=1" compiles in also 1 minute erlang logging interval.

	* aprsis.c:
	    Code refactored to put the APRS-IS communication into its own
	    fork()ed sub-process communicating via a socketpair() with the
	    main loop.   Now reconnection time with the APRS-IS server does
	    not affect functionality of the main loop.

2007-12-05  Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* ChangeLog
	   Opened for the first time.  Version 0.06.



aprx-stat.c
/* **************************************************************** *
 *                                                                  *
 *  APRX -- 2nd generation receive-only APRS-i-gate with            *
 *          minimal requirement of esoteric facilities or           *
 *          libraries of any kind beyond UNIX system libc.          *
 *                                                                  *
 * (c) Matti Aarnio - OH2MQK,  2007,2008                            *
 *                                                                  *
 * **************************************************************** */


#include "aprx.h"

time_t now;
int erlangout;
int epochtime;
char *aprxlogfile; /* linkage dummy */
const char *mycall; /* linkage dummy */

void erlang_snmp(void)
{
	int i;

	/* SNMP data output - continuously growing counters
	 */

	printf("APRX.pid     %8ld\n", (long)ErlangHead->server_pid);
	printf("APRX.uptime  %8ld\n", (long)(time(NULL) - ErlangHead->start_time));
	printf("APRX.mycall  %s\n", ErlangHead->mycall);

	for (i = 0; i < ErlangLinesCount; ++i) {
	  struct erlangline *E = ErlangLines[i];

	  printf("%s   %ld %ld   %ld  %ld     %d\n",
		 E->name,
		 E->SNMP.bytes_rx, E->SNMP.packets_rx,
		 E->SNMP.bytes_tx, E->SNMP.packets_tx,
		 (int)(now - E->last_update));
	}
}

void erlang_xml(int topmode)
{
	int i, j, k, t;

	/* What this outputs is not XML, but a mild approximation
	   of the data that XML version would output.. 
	   It is not even the whole dataset, just last 60 samples
	   of each type.
	*/


	printf("APRX.pid     %8ld\n", (long)ErlangHead->server_pid);
	printf("APRX.uptime  %8ld\n", (long)(time(NULL) - ErlangHead->start_time));
	printf("APRX.mycall  %s\n", ErlangHead->mycall);

	for (i = 0; i < ErlangLinesCount; ++i) {
	  struct erlangline *E = ErlangLines[i];
	  char logtime[40];
	  struct tm *wallclock;

	  printf("\nSNMP  %s   %ld %ld   %ld  %ld     %d\n",
		 E->name,
		 E->SNMP.bytes_rx, E->SNMP.packets_rx,
		 E->SNMP.bytes_tx, E->SNMP.packets_tx,
		 (int)(now - E->last_update));

	  printf("\n1min data\n");
	  k = E->e1_cursor;
	  t = E->e1_max; if (topmode) t = 90;
	  for (j = 0; j < t; ++j) {
	    --k; if (k < 0) k = E->e1_max-1;
	    if (E->e1[k].update == 0) continue;
	    if (epochtime) {
	      sprintf(logtime,"%ld",(long)E->e1[k].update);
	    } else {
	      wallclock = gmtime(&E->e1[k].update);
	      strftime(logtime, sizeof(logtime), "%Y-%m-%d %H:%M", wallclock);
	    }
	    printf("%s  %s %2dm  %5ld  %3ld  %5ld  %3ld   %5.3f  %5.3f\n",
		   logtime, E->name, 1,
		   E->e1[k].bytes_rx,  E->e1[k].packets_rx,
		   E->e1[k].bytes_tx,  E->e1[k].packets_tx,
		   (float)E->e1[k].bytes_rx/(float)E->erlang_capa,
		   (float)E->e1[k].bytes_tx/(float)E->erlang_capa);
	  }


	  printf("\n10min data\n");
	  k = E->e10_cursor;
	  t = E->e10_max; if (topmode) t = 10;
	  for (j = 0; j < t; ++j) {
	    --k; if (k < 0) k = E->e10_max-1;
	    if (E->e10[k].update == 0) continue;
	    if (epochtime) {
	      sprintf(logtime,"%ld",(long)E->e10[k].update);
	    } else {
	      wallclock = gmtime(&E->e10[k].update);
	      strftime(logtime, sizeof(logtime), "%Y-%m-%d %H:%M", wallclock);
	    }
	    printf("%s  %s %2dm  %5ld  %3ld  %5ld  %3ld   %5.3f  %5.3f\n",
		   logtime, E->name, 10,
		   E->e10[k].bytes_rx,  E->e10[k].packets_rx,
		   E->e10[k].bytes_tx,  E->e10[k].packets_tx,
		   (float)E->e10[k].bytes_rx/((float)E->erlang_capa*10.0),
		   (float)E->e10[k].bytes_tx/((float)E->erlang_capa*10.0));
	  }


	  printf("\n60min data\n");
	  k = E->e60_cursor;
	  t = E->e60_max; if (topmode) t = 3;
	  for (j = 0; j < t; ++j) {
	    --k; if (k < 0) k = E->e60_max-1;
	    if (E->e60[k].update == 0) continue;
	    if (epochtime) {
	      sprintf(logtime,"%ld",(long)E->e60[k].update);
	    } else {
	      wallclock = gmtime(&E->e60[k].update);
	      strftime(logtime, sizeof(logtime), "%Y-%m-%d %H:%M", wallclock);
	    }
	    printf("%s  %s %2dm  %5ld  %3ld  %5ld  %3ld   %5.3f  %5.3f\n",
		   logtime, E->name, 60,
		   E->e60[k].bytes_rx,  E->e60[k].packets_rx,
		   E->e60[k].bytes_tx,  E->e60[k].packets_tx,
		   (float)E->e60[k].bytes_rx/((float)E->erlang_capa*60.0),
		   (float)E->e60[k].bytes_tx/((float)E->erlang_capa*60.0));
	  }



	}


	exit(0);
}


void usage(void)
{
	printf("Usage: aprx-stat [-t] [-f arpx-erlang.dat] {-S|-x|-X}\n");
	exit(64);
}

int main(int argc, char **argv)
{
	int opt;
	int mode_snmp = 0;
	int mode_xml  = 0;

	now = time(NULL);

	while ((opt = getopt(argc, argv, "f:StxX?h")) != -1) {
	  switch (opt) {
	  case 'f':
	    erlang_backingstore = optarg;
	    break;
	  case 'S': /* SNMP */
	    ++mode_snmp;
	    break;
	  case 'X':
	    mode_xml = 1;
	    break;
	  case 'x':
	    mode_xml = 2;
	    break;
	  case 't':
	    epochtime = 1;
	    break;
	  default:
	    usage();
	    break;
	  }
	}

	erlang_start(0); /* Open the backing-store */

	if (!ErlangHead) exit(1);

	if (mode_snmp) {
	  erlang_snmp();
	} else if (mode_xml == 1) {
	  erlang_xml(0);
	} else if (mode_xml == 2) {
	  erlang_xml(1);
	} else
	  usage();

	return 0;
}
